src/StreamJsonRpc.Tests/ObserverMarshalingTests.cs

```diff
@@ -0,0 +1,383 @@
+    }
+
+    protected abstract IJsonRpcMessageFormatter CreateFormatter();
+
+    protected override void Dispose(bool disposing)
+    {
+        this.serverRpc.Dispose();
+        this.client.Dispose();
+
+        if (this.clientRpc.Completion.IsFaulted)
+        {
+            this.Logger.WriteLine("Client faulted with: " + this.clientRpc.Completion.Exception);
+        }
+
+        if (this.serverRpc.Completion.IsFaulted)
+        {
+            this.Logger.WriteLine("Server faulted with: " + this.serverRpc.Completion.Exception);
+        }
+
+        base.Dispose(disposing);
+    }
+
+    [MethodImpl(MethodImplOptions.NoInlining)]
+    private async Task<(WeakReference Observer, WeakReference Proxy)> CancelableSubscription_Helper()
+    {
+        var mockObserver = new MockObserver<int>();
+        var observedValueCount = 0;
+        var allValuesObserved = new AsyncManualResetEvent();
+        mockObserver.Next += (s, e) =>
+        {
+            if (++observedValueCount == 2)
+            {
+                allValuesObserved.Set();
+            }
+        };
+
+        IDisposable disposable = await this.client.Subscribe(mockObserver).WithCancellation(this.TimeoutToken);
+        await allValuesObserved.WaitAsync(this.TimeoutToken);
+        await (this.server.FireAndForgetTask ?? Task.CompletedTask).WithCancellation(this.TimeoutToken);
+        disposable.Dispose();
+
+        var weakObserver = new WeakReference(mockObserver);
+        var weakProxy = new WeakReference(this.server.Observer);
+        this.server.Observer = null;
+        return (weakObserver, weakProxy);
+    }
+
+    [MethodImpl(MethodImplOptions.NoInlining)]
+    private async Task<WeakReference> ServerMethodNotFound_ReleasesClientReference_Helper()
+    {
+        var observer = new MockObserver<int>();
+        await Assert.ThrowsAsync<RemoteMethodNotFoundException>(() => Task.Run(() => this.clientRpc.InvokeWithCancellationAsync("non-existing", new object?[] { observer }, new Type[] { typeof(IObserver<int>) }, this.TimeoutToken)).WithCancellation(this.TimeoutToken));
+        return new WeakReference(observer);
+    }
+
+    [MethodImpl(MethodImplOptions.NoInlining)]
+    private async Task<WeakReference> ServerMethodThrows_ReleasesClientReference_Helper()
+    {
+        var observer = new MockObserver<int>();
+        await Assert.ThrowsAsync<RemoteInvocationException>(() => Task.Run(() => this.client.ThrowImmediately(observer)).WithCancellation(this.TimeoutToken));
+        return new WeakReference(observer);
+    }
+
+    [MethodImpl(MethodImplOptions.NoInlining)]
+    private async Task<WeakReference> PushCompleteAndReturn_Helper()
+    {
+        var observer = new MockObserver<int>();
+        await Task.Run(() => this.client.PushCompleteAndReturn(observer)).WithCancellation(this.TimeoutToken);
+        await observer.Completion.WithCancellation(this.TimeoutToken);
+        var result = await observer.Completion;
```

Comment:
Nit: Can we not do "var result = await observer.Completion.WithCancellation(..);"

Reply:
I assume it's the `var` in particular that you dislike. Yes, I dislike it too. I've fixed it.